### PR TITLE
improvement(hydra): reduce size of the Hydra image

### PR DESCRIPTION
**Pull request by enaydanov**
Made 2 changes:
 - switched to smaller python:3.8.3-slim-buster base image
 - switched to multi-stage build

Image size reduced from 1.46GB to 618MB (compressed size reduced from 540MB to 185MB, i.e. about 3 times.)

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [x] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

---

PS: There is a good article about a base image for Python apps: https://pythonspeed.com/articles/base-image-python-docker-images/

### CODE REVIEW

File: sdcm/cluster.py [hunk 1]
```diff
@@ -3682,8 +3682,8 @@ def get_cluster_manager(self):
         mgr_cluster = manager_tool.get_cluster(cluster_name)
         if not mgr_cluster:
             self.log.debug("Could not find cluster : {} on Manager. Adding it to Manager".format(cluster_name))
-            targets = [[n, n.ip_address] for n in self.nodes]
-            mgr_cluster = manager_tool.add_cluster(name=cluster_name, host=targets[0], disable_automatic_repair=True,
+            target = self.nodes[0].ip_address
```
Comment (fruch) [inline sdcm/cluster.py:3685]: can you move it it into it's own commit or PR ?

Comment (enaydanov) [reply sdcm/cluster.py:3685]: Yep. I just try to pass thru provision test here.

Comment (enaydanov) [reply sdcm/cluster.py:3685]: Done: https://github.com/scylladb/scylla-cluster-tests/pull/2356

**Comment by enaydanov** [comment]
All provision tests are passed actually. Also, it looks like 1h is not enough for them.

**Comment by fruch** [comment]
> All provision tests are passed actually. Also, it looks like 1h is not enough for them.

I would recommend put one hour on each stage.

In general putting the timeout on the whole pipeline is problematic, since it would include waiting time for the work to be available. ( Most time we have enough, but not always)

File: docker/env/Dockerfile [hunk 2]
```diff
@@ -1,18 +1,42 @@
...
+RUN apt-get install -y --no-install-recommends build-essential
+ADD requirements-python.txt .
+RUN mkdir /build && \
+    pip3 install -r requirements-python.txt --root=/build --prefix=./ --ignore-installed --no-warn-script-location
 
-ENV PYTHONWARNINGS ignore:unclosed
+FROM python:$PYTHON_IMAGE_TAG
+ENV PYTHONWARNINGS=ignore:unclosed \
+    PYTHONFAULTHANDLER=1 \
+    PYTHONHASHSEED=random \
```
Comment (fruch) [inline docker/env/Dockerfile:24]: Why are we changing this default ?

File: docker/env/Dockerfile [hunk 3]
```diff
@@ -1,18 +1,42 @@
...
+        git \
+        iproute2 \
+        openssh-client \
+        rsync \
+        sudo \
+        unzip \
+        wget \
+        docker-ce-cli && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (fruch) [inline docker/env/Dockerfile:41]: For debugging sake, maybe we should leave the lists ? 
I.e. when you are trying to do apt install, you then need to lookup the list and out them into place

Comment (fruch) [reply docker/env/Dockerfile:41]: Never mind I've got it wrong

File: docker/env/Dockerfile [hunk 4]
```diff
@@ -1,18 +1,42 @@
...
+RUN apt-get update
 
-# Install Docker cli only
-RUN curl -fsSL https://download.docker.com/linux/debian/gpg | sudo apt-key add - && \
-    DEBIAN_FRONTEND=noninteractive add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable" && \
-    DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install -y docker-ce-cli  && \
-    sudo groupadd docker && \
-    sudo usermod -aG docker `whoami`
+# Add 3rd-party APT repositories.
+FROM apt_base as apt_repos
```
Comment (fruch) [inline docker/env/Dockerfile:8]: Why doing this part in a separate build stage ? Seem a bit overkill, and you end up installing curl twice cause of that.

The three packages here are that big of difference in image size ?

Comment (enaydanov) [reply docker/env/Dockerfile:8]: The cause that `software-properties-common` (and its friends) has a lot of dependencies which we don't need:

```
$ apt install --no-install-recommends --dry-run apt-transport-https gnupg2 software-properties-common
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  dirmngr distro-info-data gir1.2-glib-2.0 gir1.2-packagekitglib-1.0 gnupg gnupg-l10n gnupg-utils gpg gpg-agent gpg-wks-client gpg-wks-server gpgconf gpgsm iso-codes libapt-inst2.0 libassuan0 libcurl3-gnutls libdbus-1-3 libgirepository-1.0-1 libglib2.0-0
  libgssapi-krb5-2 libk5crypto3 libkeyutils1 libkrb5-3 libkrb5support0 libksba8 libldap-2.4-2 libldap-common libmpdec2 libnghttp2-14 libnpth0 libpackagekit-glib2-18 libpsl5 libpython3-stdlib libpython3.7-minimal libpython3.7-stdlib librtmp1 libsasl2-2
  libsasl2-modules-db libssh2-1 lsb-base lsb-release mime-support pinentry-curses python-apt-common python3 python3-apt python3-dbus python3-gi python3-minimal python3-pycurl python3-software-properties python3.7 python3.7-minimal
Suggested packages:
  dbus-user-session libpam-systemd pinentry-gnome3 tor parcimonie xloadimage scdaemon isoquery krb5-doc krb5-user lsb pinentry-doc python3-doc python3-tk python3-venv python3-apt-dbg python-apt-doc python-dbus-doc python3-dbus-dbg libcurl4-gnutls-dev python-pycurl-doc
  python3-pycurl-dbg python3.7-venv python3.7-doc binutils binfmt-support
Recommended packages:
  dbus libglib2.0-data shared-mime-info xdg-user-dirs krb5-locales packagekit publicsuffix libsasl2-modules bzip2 file xz-utils unattended-upgrades
The following NEW packages will be installed:
  apt-transport-https dirmngr distro-info-data gir1.2-glib-2.0 gir1.2-packagekitglib-1.0 gnupg gnupg-l10n gnupg-utils gnupg2 gpg gpg-agent gpg-wks-client gpg-wks-server gpgconf gpgsm iso-codes libapt-inst2.0 libassuan0 libcurl3-gnutls libdbus-1-3 libgirepository-1.0-1
  libglib2.0-0 libgssapi-krb5-2 libk5crypto3 libkeyutils1 libkrb5-3 libkrb5support0 libksba8 libldap-2.4-2 libldap-common libmpdec2 libnghttp2-14 libnpth0 libpackagekit-glib2-18 libpsl5 libpython3-stdlib libpython3.7-minimal libpython3.7-stdlib librtmp1 libsasl2-2
  libsasl2-modules-db libssh2-1 lsb-base lsb-release mime-support pinentry-curses python-apt-common python3 python3-apt python3-dbus python3-gi python3-minimal python3-pycurl python3-software-properties python3.7 python3.7-minimal software-properties-common
0 upgraded, 57 newly installed, 0 to remove and 0 not upgraded.
```

So, yes, these 3 images add about 70MB to the image.

File: docker/env/Dockerfile (hunk 2 above)
Comment (enaydanov) [reply docker/env/Dockerfile:24]: Just copied this stuff from recommended one (I think, it was python-poetry stuff.)
But due to https://docs.python.org/3.4/using/cmdline.html#envvar-PYTHONHASHSEED `random` is a default value.

I'll remove it.

**Comment by enaydanov** [comment]
Also we can cut another 74MB (24MB compressed) if we'll skip git installation. `git` binary used twice inside hydra: as a duplicate of a code in `hydra.sh` for getting git user email (`sdcm/utils/get_username.py`). Another one is a getting SCT commit hash in `sdcm/db_stats.py` which we can pass to the hydra via env variable in `hydra.sh` and/or we can use Docker version of git (e.g., [alpine/git](https://hub.docker.com/r/alpine/git/))
